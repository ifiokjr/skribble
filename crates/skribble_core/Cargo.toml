[package]
name = "skribble_core"
version = { workspace = true }
authors = { workspace = true }
categories = { workspace = true }
documentation = "https://docs.rs/skribble_core"
edition = { workspace = true }
homepage = { workspace = true }
include = ["src/**/*.rs", "Cargo.toml", "readme.md"]
keywords = ["build", "atomic", "css"]
license = { workspace = true }
readme = "readme.md"
repository = { workspace = true }
rust-version = { workspace = true }
description = "Core features for the skribble atomic css library"

[dependencies]
derivative = { workspace = true }
derive_more = { workspace = true }
doc-comment = { workspace = true }
globset = { workspace = true }
heck = { workspace = true }
indent_write = { workspace = true }
indexmap = { features = ["serde"], workspace = true }
lazy_static = { workspace = true }
lightningcss = { workspace = true }
readonly = { workspace = true }
regex = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
skribble_color = { workspace = true, features = ["serde"] }
thiserror = { workspace = true }
typed-builder = { workspace = true }
vfs = { workspace = true }

[dev-dependencies]
insta = { features = ["json"], workspace = true }
