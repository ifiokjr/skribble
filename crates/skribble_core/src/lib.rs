#![deny(clippy::all)]

pub use error::*;

mod error;
