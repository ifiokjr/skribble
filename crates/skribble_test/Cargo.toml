[package]
name = "skribble_test"
version = { workspace = true }
authors = ["Ifiok Jr. <ifiokotung@gmail.com>"]
documentation = "https://docs.rs/skribble_test"
edition = { workspace = true }
homepage = { workspace = true }
include = ["src/**/*.rs", "Cargo.toml", "readme.md"]
keywords = ["build", "atomic", "css"]
license = { workspace = true }
publish = false
readme = "readme.md"
repository = { workspace = true }
rust-version = { workspace = true }
description = "internal test helpers for skribble"

[dependencies]

[dev-dependencies]
insta = { workspace = true }
rstest = { workspace = true }
similar-asserts = { workspace = true }
