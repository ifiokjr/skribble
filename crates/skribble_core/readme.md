# skribble_core

> core library for skribble

<br />

[![Crate][crate-image]][crate-link] [![Docs][docs-image]][docs-link] [![Status][ci-status-image]][ci-status-link] [![Unlicense][unlicense-image]][unlicense-link]

## Installation

```toml
[dependencies]
skribble_core = "0.0.0"
```

### Usage

```rust
use skribble_core::*;
```

[crate-image]: https://img.shields.io/crates/v/skribble_core.svg
[crate-link]: https://crates.io/crates/skribble_core
[docs-image]: https://docs.rs/skribble_core/badge.svg
[docs-link]: https://docs.rs/skribble_core
[ci-status-image]: https://github.com/ifiokjr/skribble/workflows/ci/badge.svg
[ci-status-link]: https://github.com/ifiokjr/skribble/actions?query=workflow:ci
[unlicense-image]: https://img.shields.io/badge/license-Unlicence-blue.svg
[unlicense-link]: https://opensource.org/license/unlicense
