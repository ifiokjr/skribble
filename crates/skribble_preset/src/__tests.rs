use rstest::*;
use skribble_core::vfs::MemoryFS;
use skribble_core::*;
use skribble_test::set_snapshot_suffix;

use super::*;

#[test]
fn default_can_be_added_to_runner() -> AnyEmptyResult {
  let plugin = PresetPlugin::builder().reset("tailwindCompat").build();
  let config: StyleConfig = StyleConfig::builder()
    .plugins(vec![PluginContainer::from(plugin)])
    .build();

  let mut runner = SkribbleRunner::try_new(config)?;
  let runner_config = runner.initialize()?;
  insta::assert_json_snapshot!(runner_config);

  Ok(())
}

#[rstest]
#[case("contained", &["$contained", "lg:$contained", "xl:$contained"])]
#[case("aspect-ratio", &["aspect:$square", "aspect:$portrait", "aspect:[2/1]"])]
fn css_from_class_names(#[case] id: &str, #[case] names: &[&str]) -> AnyEmptyResult {
  let plugin = PresetPlugin::default();
  let config: StyleConfig = StyleConfig::builder()
    .plugins(vec![PluginContainer::from(plugin)])
    .build();
  let mut runner = SkribbleRunner::try_new(config)?;
  let runner_config = runner.initialize()?;
  let mut classes = Classes::default();

  for name in names.iter() {
    classes.insert_factory(ClassFactory::from_string(runner_config, name));
  }

  classes.sort_by_class();

  set_snapshot_suffix!("{id}");
  insta::assert_display_snapshot!(classes.to_skribble_css(runner_config)?);

  Ok(())
}

#[test]
fn auto_generate_reset_css() -> AnyEmptyResult {
  let plugin = PresetPlugin::builder().reset("tailwindCompat").build();
  let config: StyleConfig = StyleConfig::builder()
    .plugins(vec![PluginContainer::from(plugin)])
    .build();

  let vfs: VfsPath = create_memory_fs()?;
  let mut runner = SkribbleRunner::new(config, "/", Some(vfs));
  let _ = runner.initialize()?;
  let scanned = runner.scan()?;
  insta::assert_display_snapshot!(scanned.code);

  Ok(())
}

fn create_memory_fs() -> AnyResult<VfsPath> {
  let vfs: VfsPath = MemoryFS::new().into();

  for file in FILES {
    let path = vfs.join(file.0)?;
    path.create_dir_all()?;
    let mut writer = path.create_file()?;
    write!(writer, "{}", file.1)?;
  }

  Ok(vfs)
}

const FILES: &[(&str, &str)] = &[];
