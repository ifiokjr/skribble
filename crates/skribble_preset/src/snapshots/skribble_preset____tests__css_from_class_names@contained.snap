---
source: crates/skribble_preset/src/__tests.rs
expression: classes.to_skribble_css(runner_config)?
---
/* Generated by Skribble */
@charset "utf-8";
@layer base, default;
@property --sk-cmw {
  syntax: "*";
  inherits: true;
  initial-value: inherit;
}
@layer default {
  .\$contained {
    width: 100%;
    max-width: var(--sk-cmw);
  }
  @media (min-width: 640px) {
    :root {
      --sk-cmw: 640px;
    }
  }
  @media (min-width: 768px) {
    :root {
      --sk-cmw: 768px;
    }
  }
  @media (min-width: 1024px) {
    :root {
      --sk-cmw: 1024px;
    }
    .lg\:\$contained {
      width: 100%;
      max-width: var(--sk-cmw);
    }
  }
  @media (min-width: 1280px) {
    :root {
      --sk-cmw: 1280px;
    }
    .xl\:\$contained {
      width: 100%;
      max-width: var(--sk-cmw);
    }
  }
  @media (min-width: 1536px) {
    :root {
      --sk-cmw: 1536px;
    }
  }
}

